src/persist-client/src/examples/persistent_source_example.rs

```diff
@@ -0,0 +1,1395 @@
+                inclusive_upper,
+            };
+
+            let message = Message {
+                capability: message.capability,
+                payload: timestamped_batch,
+            };
+            if let Err(_) = tx.send(message).await {
+                error!("receiver dropped");
+                return;
+            }
+        }
+    }
+
+    /// The egress point of the source pipeline. This could either write into persistence or feed
+    /// straight into a timely dataflow.
+    async fn emit<D, TS, FT, TT>(
+        name: String,
+        timestamper: Arc<Mutex<TS>>,
+        source_frontier: Arc<Mutex<MutableAntichain<FT>>>,
+        mut rx: mpsc::Receiver<Message<FT, TimestampedBatch<D, TT>>>,
+        mut write: WriteHandle<D, (), TS::ToTimestamp, i64>,
+    ) where
+        D: Codec + Debug + Send + Sync,
+        TS: Timestamper<FromTimestamp = FT, ToTimestamp = TT>,
+        FT: timely::PartialOrder + PartialOrd + Ord + Send + Sync + Debug + Copy,
+        TT: timely::progress::Timestamp
+            + Lattice
+            + Codec64
+            + timely::PartialOrder
+            + Ord
+            + Debug
+            + Copy,
+    {
+        // Figure out what everyone else has written so far, and don't write updates that are not
+        // beyond that.
+        let mut initial_upper = write.upper().clone();
+
+        // We need to stash batches whose upper is beyond the frontier, so that we can "atomically"
+        // write all batches and downgrade to a new upper when the frontier advances.
+        let mut stashed_batches = Vec::new();
+
+        while let Some(mut message) = rx.recv().await {
+            let batch = message.payload;
+            println!("instance {}: emit got = {:?}", name, batch);
+
+            let capability_subtraction = message.capability.drain().map(|(ts, diff)| (ts, -diff));
+            source_frontier
+                .lock()
+                .await
+                .update_iter(capability_subtraction);
+
+            let source_frontier = source_frontier.lock().await.frontier().to_owned();
+
+            println!("instance {}: source_frontier: {:?}", name, source_frontier,);
+
+            let all_bindings = timestamper.lock().await.all_bindings().await;
+
+            let open_timestamps = all_bindings
+                .iter()
+                .filter(|(_to, from)| source_frontier.less_than(&from))
+                .map(|(to, _from)| *to);
+
+            let mut target_upper = Antichain::new();
+            target_upper.extend(open_timestamps);
+
+            let largest_target_ts = all_bindings
+                .iter()
+                .map(|(to, _from)| *to)
+                .max()
+                .expect("there must be some bindings");
+
+            // It's **very** important that we also insert the largest seen timestamp because the
+            // open timestamps might be empty. And then we would be downgrading to the empty
+            // frontier, which ...not good!
```

Comment:
nit: s/...not good!/closes the shard permanently for writes./

Reply:
👍